src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs

```diff
@@ -45,76 +42,96 @@ private static async ValueTask<TValue> ReadAsync<TValue>(Stream utf8Json, Type r
+            int bytesInBuffer = 0;
+            long totalBytesRead = 0;
+            int clearMax = 0;
 
             try
             {
-                do
+                while (true)
                 {
-                    int bytesToRead = bufferSize - bytesRemaining;
-                    bytesRead = await utf8Json.ReadAsync(buffer, bytesRemaining, bytesToRead, cancellationToken).ConfigureAwait(false);
+                    // Read from the stream until either our buffer is filled or we hit EOF.
+                    // Calling ReadCore is relatively expensive, so we minimize the number of times
+                    // we need to call it.
+                    bool isFinalBlock = false;
+                    while (true)
+                    {
+                        int bytesRead = await utf8Json.ReadAsync(buffer.AsMemory(bytesInBuffer), cancellationToken).ConfigureAwait(false);
+
+                        if (bytesRead == 0)
+                        {
+                            isFinalBlock = true;
+                            break;
+                        }
 
-                    deserializeBufferSize = bytesRemaining + bytesRead;
-                    isFinalBlock = (bytesRead == 0);
+                        totalBytesRead += bytesRead;
+                        bytesInBuffer += bytesRead;
 
+                        if (bytesInBuffer > clearMax)
+                        {
+                            clearMax = bytesRead;
+                        }
+
+                        if (bytesInBuffer == buffer.Length)
+                        {
+                            break;
+                        }
+                    }
+
+                    // Process the data available
                     ReadCore(
                         ref readerState,
                         isFinalBlock,
-                        buffer,
-                        deserializeBufferSize,
+                        new Span<byte>(buffer, 0, bytesInBuffer),
                         options,
                         ref state);
 
                     int bytesConsumed = (int)readerState.BytesConsumed;
-                    bytesRemaining = deserializeBufferSize - bytesConsumed;
+                    bytesInBuffer -= bytesConsumed;
 
                     if (isFinalBlock)
                     {
                         break;
                     }
 
                     // Check if we need to shift or expand the buffer because there wasn't enough data to complete deserialization.
-                    if (bytesConsumed <= (bufferSize / 2))
+                    if ((uint)bytesInBuffer > ((uint)buffer.Length / 2))
                     {
                         // We have less than half the buffer available, double the buffer size.
-                        bufferSize = (bufferSize < HalfMaxValue) ? bufferSize * 2 : int.MaxValue;
+                        byte[] dest = ArrayPool<byte>.Shared.Rent((buffer.Length < (int.MaxValue / 2)) ? buffer.Length * 2 : int.MaxValue);
```

Comment:
I asked this before (https://github.com/dotnet/corefx/pull/35609#discussion_r261813289), but

> Renting `int.MaxValue` would throw OOM. Should we cap it to something smaller (either 2GB, or maybe `int.MaxValue - 56`)?

Is there guidelines or common pattern to follow in such scenarios?